### ISSUE TITLE
kops doesn't tag a number of resources it creates

### ISSUE THREAD

**Issue by s3asfour** [kubernetes/kops#3358]
At my company we're required to tag all the resources in AWS so that we can track resources related to a project and get a better insight into billing. I was tasked with adding tags to all resources that kops create, but after looking into it it seems to me that kops allows us to create EC2 instances only, and I can tag ELBs using kubernetes services file annotations. But what about volumes, network interfaces and security groups that kops create?

Basically i'd like kops to tag every AWS resources that allows tagging with the same tags.

**Comment by chrislovecnm** [comment]
Aws does not support the tagging of volumes created by asg. You would need to use lambda for that. Security groups should be tagged. ENI and bastions I think are tagged as well. Can you do a list of what is not tagged with kops 1.7.1 alpha?

**Comment by s3asfour** [comment]
I am using kops version v1.6.2, which should have cloud labels support to tag the resources it creates in AWS.

The following resources are not tagged by kops:

- Security groups
- EBS Volumes
- Network interfaces

**Comment by chrislovecnm** [comment]
Need to check if kops 1.7.x us tagging security groups. The root ebs volumes and nics cannot be tagged by kops at this time.  As Das as I know AWS does not provide the capability for an asg to label Ebs and nics created by the asg. You can use lambda functions or other post processing programs to label.

**Comment by philipbjorge** [comment]
Kops 1.7.1 tagged my security groups.

**Comment by chrislovecnm** [comment]
Marked this as good starter issue.

1. determine if we can tag any other resources
2. write the code to tag them
3. if we cannot tag certain resources, such as disks created from ASGs, then document that

**Comment by chrislovecnm** [comment]
Some work on this: https://github.com/kubernetes/kops/pull/4489

**Comment by chrislovecnm** [comment]
> Security groups

Are tagged.

We cannot tag EBS volumes that are created by ASGs and they need to be tagged after the instance is spun up.

**Comment by mazzy89** [comment]
I use the following hook to assign tags to EBS volumes:

```
    - name: kops-hook-tag-ebs-volumes.service
      roles:
        - Master
        - Node
      manifest: |-
        [Unit]
        Description=Tag EBS Volumes without tags with AutoScaling Group tags
        [Service]
        Type=oneshot
        ExecStartPre=/bin/bash -c "/usr/bin/curl -s https://stedolan.github.io/jq/download/linux64/jq > /usr/local/bin/jq && chmod +x /usr/local/bin/jq"
        ExecStart=/bin/bash \
          -c 'AWS_REGION=$(/usr/bin/curl -s http://169.254.169.254/latest/meta-data/placement/availability-zone | sed 's/[a-z]$//'); \
              INSTANCE_ID=$(/usr/bin/curl -s http://169.254.169.254/latest/meta-data/instance-id); \
              VOLUMES=$(/usr/local/bin/aws ec2 describe-volumes --region $AWS_REGION --filters Name=attachment.instance-id,Values=$INSTANCE_ID | /usr/local/bin/jq -r '"'"'.Volumes[] | select(.Tags == null) | .Attachments[].VolumeId'"'"'); \
              AUTOSCALING_GROUP=$(/usr/local/bin/aws autoscaling describe-auto-scaling-instances --region $AWS_REGION --instance-ids $INSTANCE_ID | /usr/local/bin/jq -r .AutoScalingInstances[].AutoScalingGroupName); \
              TAGS=$(/usr/local/bin/aws autoscaling describe-tags --region $AWS_REGION --filters Name=auto-scaling-group,Values=$AUTOSCALING_GROUP --query '"'"'Tags[*].{Key:Key,Value:Value}'"'"'); \
              /usr/local/bin/aws ec2 create-tags --region "$AWS_REGION" --resources "$VOLUMES" --tags "$TAGS";'
```

However it doesn't work for the bastions ASG because it doesn't use the `kubeup`

**Comment by William-Luo0** [comment]
EBS volumes that are spun up with launch templates can be tagged including root volumes. Launch templates also have other advantages over launch config so it may be beneficial to move towards launch templates.

**Comment by mikesplain** [comment]
FYI @alex-hempel this should be fixed in https://github.com/kubernetes/kops/pull/6703 which will be in Kops 1.12

**Comment by techdragon** [comment]
EBS Volumes tagged could be supported by switching over to only using Autoscaling groups with Launch Templates. - https://forums.aws.amazon.com/thread.jspa?threadID=122354&start=25&tstart=0

**Comment by hakman** [comment]
@techdragon I was looking at the same thing. Seem we use Launch Templates already when `mixedInstancesPolicy` is used, but not in general. Not sure why though.

**Comment by hakman** [comment]
@gambol99 I see that you introduced Launch Templates in https://github.com/kubernetes/kops/pull/6277. Any reason we don't use them in general, not just for Mixed Instances Policy?

**Comment by AndreKapraty** [comment]
Is there any chance that It will be implemented? We use some user data scripts for resource tagging, but It's not helpful for etcd volumes. Every time when we create new IG and execute "update cluster" It will re-tag etcd volumes. We need that because we have several accounts with billing etc and so, we tag everything.

**Comment by csmartins** [comment]
Hello, is there any chance of this ever getting implemented? I would mostly like to have ENIs that are attached to the instances receiving the same tags as them.

**Comment by hakman** [comment]
> Hello, is there any chance of this ever getting implemented? I would mostly like to have ENIs that are attached to the instances receiving the same tags as them.

@csmartins I think this is possible now. I will give it a try in #17773.

Many of the missing tags were addressed with or without referencing this issue over the past years, when AWS added support. If any resources remain untagged, those interested should create separate issues for them, for easy tracking.
